### ISSUE TITLE
DTO Image

### ISSUE THREAD

**Issue by pkriens** [pkriens/cdi#34]
![image](https://user-images.githubusercontent.com/200494/32775797-c9c1aca8-c930-11e7-8213-9442fa4d7be8.png)

**Comment by pkriens** [comment]
// model
enum  boolean { true, false }

sig PID {}
sig ServiceID, BundleID, Version {}
sig ExtensionID, ContainerID, BeanID, TemplateID {}
one sig OSGiExtensionID extends ExtensionID {}

enum ServiceScope { NO_SERVICE, SINGLETON, BUNDLE, PROTOTYPE }
sig Class {}
one sig Object extends Class {}
one sig CDIApplicationClass extends Class {}

sig Map {
	service_pid : PID
}
abstract sig DTO {}
sig Filter {}
sig ServiceReferenceDTO extends DTO {}
sig BundleDTO extends DTO {}

enum ContainerState {
	CREATING,
	CREATED,
	DESTROYING,
	DESTROYED,
	WAITING_FOR_REFERENCES,
	WAITING_FOR_EXTENSIONS,
	SATISFIED,
	FAILURE
}

one sig ContainerDTO extends DTO {
	id					 						: ContainerID,
	state										: ContainerState,
	extensions 							: some ExtensionDTO,
	application							: lone ComponentInstanceDTO,
	templates								: set ComponentTemplateDTO,
	instances								: set ComponentInstanceDTO,
	bundle									: BundleID
} {
	OSGiExtensionID in extensions.id

	all disj e1, e2 : extensions | e1.id != e2.id
	all disj t1, t2 : templates | t1.id != t2.id

	all i : instances.id | i in templates.id

	state not in SATISFIED implies no instances
 
	state = SATISFIED implies 
		extensions.satisfied = true
	and
		one application
	and {
		application.satisfied = true 
		one application.serviceId
	}

	one application implies {
		application.serviceScope = SINGLETON
		application.objectClass = CDIApplicationClass
		application.configurationFactory = false
		// application.lifecyclePID = ???
		application.configurationFactory = false
		no application.extraPIDs
		application.enabled = true
	}

	state in WAITING_FOR_REFERENCES implies extensions.satisfied = true
}

sig ExtensionDTO extends DTO {
	id											: ExtensionID,
	version									: Version,
	service									: lone ServiceID,
	satisfied								: boolean
} {
	satisfied = true <=> one service
}

sig ComponentTemplateDTO extends DTO {
	id											: TemplateID,
	lifecyclePID						: PID,
	configurationFactory		: boolean,
	extraPIDs								: set ConfigurationTemplateDTO,
	serviceTemplates	 			: set ServiceTemplateDTO,
	configurationTemplates 	: set ConfigurationTemplateDTO,
	enabled									: boolean,
	serviceScope						: ServiceScope,
	objectClass							: set Class
} {
	serviceScope = NO_SERVICE implies no objectClass else some objectClass

	enabled = false implies no i : ContainerDTO.instances | i.@id = id

	extraPIDs in configurationTemplates
}

sig ServiceTemplateDTO extends DTO {
	id											: TemplateID,
	objectClass							: Class,
	targetFilter						: lone Filter,
	maximumCardinality			: Int,
	minimumCardinality			: Int,
	greedy									: boolean,
	dynamic									: boolean,
	referenceBean						: BeanID
} {
	minimumCardinality >= 0
	maximumCardinality <= 8
}

sig ConfigurationTemplateDTO extends DTO {
	id											: TemplateID,
	pid											:	lone PID,
	required								: boolean
}

sig ComponentInstanceDTO extends ComponentTemplateDTO {
	satisfied								: boolean,
	configuration 					: lone Map,
	serviceDependencies			: set ServiceDependencyDTO,
	configurationDependencies: set ConfigurationDependencyDTO,
	serviceId								: lone ServiceID
} {
	((all s : serviceDependencies | s.@satisfied = true) 
		and 
	(all c : configurationDependencies | c.@satisfied = true)) 
		implies satisfied = true else satisfied = false

	this != ContainerDTO.application implies one t : ContainerDTO.templates |  t.@id = id 
}

sig ServiceDependencyDTO extends ServiceTemplateDTO {
	satisfied								: boolean,
	services								: set ServiceID
} {
	satisfied = false implies no services
	satisfied = true implies # services >= minimumCardinality and # services <= maximumCardinality
}

sig ConfigurationDependencyDTO extends ConfigurationTemplateDTO {
	satisfied								: boolean,
	configuration						: Map,
} {
	satisfied = true implies {
		some configuration
		no pid implies configuration = ContainerDTO.application.@configuration
	}
}




sig Bean {
	types						: some Class,
	injectionPoints : set InjectionPoint,
	scope 					: Scope
} {
	not this in this.^(@injectionPoints.bean) // no cycles

	Object in types
}

sig ReferenceBean extends Bean {} {
	no injectionPoints
	scope = Synthetic
}

enum Scope { Dependent, ComponentScoped, ApplicationScoped, RequestScoped, SessionScoped, Synthetic, CustomScoped }


sig InjectionPoint {
	bean			 			: Bean
} {
	one b : Bean | this in b.injectionPoints
}

sig CDIComponent extends Bean {
	references : set ReferenceBean
} {
	no ip : InjectionPoint | ip.bean = this 
	scope = ComponentScoped
	this 	in ApplicationComponent.components
}

pred isReachable[ src, dst : Bean, allowedScopes : set Scope ] {
	dst in src.injectionPoints.bean
	or some  intermediate : src.injectionPoints.bean | 
		intermediate.scope in allowedScopes 
		and isReachable[ intermediate, dst, allowedScopes ]
}

one sig ApplicationComponent {
	container 	: set Bean,
	references	: set ReferenceBean,
	components 	: set CDIComponent
} {

	Bean = container
	components = CDIComponent

	all r : ReferenceBean | 
    r in references 
		<=> 
			some b : Bean | 
				not b.scope in ComponentScoped+Dependent  
				and isReachable[b,r, ComponentScoped+Dependent ]
 
	all c : components |
		all r : ReferenceBean | 
			r in c.references 
			<=> 
				not r in references 			
				and isReachable[c,r, ComponentScoped+Dependent ] 

	one ContainerDTO.application implies {
	}
}

fact MakeComplete {


		all r : ReferenceBean | some ip : InjectionPoint | ip.bean = r
		
		all c : Bean - CDIComponent | c.scope in ComponentScoped+Dependent implies some b : Bean | c = b.injectionPoints.bean 		

}

assert References {
	no r : ApplicationComponent.references | r in CDIComponent.references
}

assert Used {
	all r : ReferenceBean | r in ApplicationComponent.references or r in CDIComponent.references
}

check References for 5

pred show {
	some ApplicationComponent.references
	some CDIComponent.references
}
run show for 4

**Comment by rinswind** [comment]
The "component lifecycle" seems to be missing.

How is a factory configuration dependency expressed?
- The produced ComponentInstances do not have a ConfigurationDependency for the factory at all?
- The produced ComponentInstances start with a ConfigurationDependency already bound to a concrete factory configuration object?
- What about the case when there are many singleton configurations and one factory configuration, and some singleton configurations are missing? Each ComponentInstance produced from the factory configuration has the same set of unsatisfied singleton ConfigurationDependency?

**Comment by pkriens** [comment]
I do not think we need a special DTO for the component life cycle since there is a 1:1 relation between the type of life cycle and the template. So we can put the lifecycle details in the ComponentTemplate.

a) Not sure what you mean? ComponentInstanceDTO has a set of ConfigurationDependency. It does not have a dependency for the lifecycle pid (either singleton or factory) since that outside the span of the component.

b) Nope, the lifecycle configuration is outside the scope of the component. I.e. the model is not wait for it, but the lifecycle configuration pushes. The other configuration dependencies are 'wait for'.

c) Again, the lifecycle pid is special and not a dependency. The extra PIDs are just dependencies. This is implied with the fact that the `extraPIDs` must be in the list of configuration templates.

Again, it is a proposal to simplify the current DTOs.

**Comment by rinswind** [comment]
I don't understand alloy script.

I can speculate that:
- Every `ComponentInstance` has 0..n `ConfigurationDependency`
- A `ConfiguratinDependency` handles one **singleton** PID.
- Every `ComponentInstance` has 0..1 `lifecyclePID`
    - Only if the component has a factory configuration.
- There is no dedicated DTO for the factory configuration

If further seems that:
```
@FactoryPID("factory")
@PID("singleton.1")
...
@PID("singleton.n")
@Component
class Foo {
    @Reference Ref1 ref1;
    ...
    @Reference RefN refN;
}
```
Will lead to:
- One `ComponentInstance` for every factory configuration with pid `factory`
- Each `ComponentInstance`
    - Will have a `ConfigurationDependency` for each `singleton.1` to `singleton.n`
    - Will have `lifecyclePid=factory`
    - Will have a `configuration` Map containing the factory configuration.
    - Will have a `ServiceDependency` for each `Ref1` to `RefN`
- For each `ComponentInstance`
    - When all `ConfigurationDependency` become `satisfied=true`
    - Each `ConfigurationDependency` will obtain a `configuration` Map
    - All `ServiceDependency` will obtain a resolved target filter
    - When all `ServiceDepdendency` become `satisfied=true`
    - Each `ServiceDependency` will have a list of `ServiceID` for the consumed services.
    - The parent `ComponentInstance` becomes `satisfied=true`

Is this all correct?
It is not clear how the "runtime" set of DTOs unfolds at runtime.

**Comment by rotty3000** [comment]
@pkriens I don't understand if this means the component must have a service id? Or doesn't this imply optional serviceId. I suspected the `lone` keyword implied "required".
```
    and {
        application.satisfied = true 
        one application.serviceId
    }
```
